gateserver/controller_protocol.py

```diff
@@ -0,0 +1,83 @@
+"""Implements the Controller <-> Server protocol.
+
+See https://github.com/fmfi-svt-gate/server/wiki/Controller-%E2%86%94-Server-Protocol .
+"""
+
+from .utils.structparse import *
+import nacl.raw as nacl
+
+PROTOCOL_VERSION = bytes([0x00,0x01])
+
+class BadMessageError(Exception): pass
+
+def checkmsg(expression, errmsg):
+    if not expression: raise BadMessageError(errmsg)
+
+class MsgType(Enum):
+    OPEN = 1
+
+class ReplyStatus(Enum):
+    OK        = 0x01
+    ERR       = 0x10
+    TRY_AGAIN = 0x11
+S = ReplyStatus
+
+PacketHead = mystruct('PacketHead',
+                     ['protocol_version', 'controllerID', 'nonce'      ],
+                     [ t.bytes(2)       ,  t.bytes(6)   ,  t.bytes(18) ])
+
+RequestHead = mystruct('RequestHead',
+                      ['msg_type'],
+                      [ t.uint8  ])
+
+ReplyHead = mystruct('ReplyHead',
+                    ['msg_type', 'status' ],
+                    [ t.uint8  ,  t.uint8 ])
+
+def crypto_unwrap(packet_head, key, payload):
+    return nacl.crypto_secretbox_open(payload,
+        packet_head.controllerID + packet_head.nonce, key)
+
+def crypto_wrap(packet_head, key, payload):
+    return nacl.crypto_secretbox(payload,
+        packet_head.controllerID + packet_head.nonce, key)
+
+def parse_packet_head(buf):
+    """Parses the packet header, returning that and the rest of the data."""
+    p, payload = PacketHead.unpack_from(buf)
+    checkmsg(p.protocol_version == PROTOCOL_VERSION, 'Invalid protocol version')
+    return p, payload
+
+def parse_r(struct, packet_head, key, payload):
```

Comment:
No, ono to pôvodne boli samostatné `parse_request` a `parse_reply`. Ale rozdiel v nich bol ten jeden riadok či `RequestHead` alebo `ReplyHead`... Mám to vrátiť?